src/rcc.rs

```diff
@@ -333,12 +335,12 @@ impl CFGR {
     ///
     /// Will result in a hang if an external oscillator is not connected or it fails to start,
     /// unless [css](CFGR::enable_css) is enabled.
-    pub fn use_hse<F>(mut self, freq: F) -> Self
+    pub fn use_hse<F>(mut self, freq: F) -> Result<Self, <F as TryInto<Hertz<u32>>>::Error>
```

Comment:
Same here, with `Result` breaking the API for a conversion error, which could be done by the user themselves, before passing it into the function. 

I'd like to reserve the change to using `Result` types to actual logic validation checks at runtime, which **can't** possibly be done manually beforehand.

Reply:
OK, so you'd prefer the functions taking a generic `F` type which are constrained to `Rate + TryInto<u32>` instead be changed to taking a `Hertz` concrete type instead?

In other words, put the onus of the conversion on the user.

So, for example:

This signature:

    pub fn pclk2<F>(mut self, freq: F) -> Result<Self, <F as TryInto<Hertz<u32>>>::Error>
    where
        F: Rate + TryInto<Hertz<u32>>;

Becomes this:

    pub fn pclk2( mut self, freq: Hertz ) -> Self;

Is this what you had in mind?

Reply:
Yes, this is what I meant. 

Maybe I'm overlooking something, but I can find no strong argument to introduce a conversion, which can fail inside the function, which error has to be propagated out of it, if the alternative is to just check the failing conversion before passing the argument into the function. 

I know, this has slight ergonomically disadvantages, but I don't think it is too bad. 
Anyways, there is no case, where the conversion into `Hertz` can beforehand, am I correct?

Reply:
Yeah that's totally fine. I had only done that initially to take the burden of the conversion off the user.
And yes, you're correct.

There is another approach as well; I figure I'll throw it out there.

We know that the maximum MCU frequency for the F3 is `74MHz` and we also know that that will comfortably fit in a `u32`. What if instead of returning a `Result` we `panic`? In the case where someone passes in a value that cannot fit into a `u32` they obviously did something wrong as no F3 as far as I know has a clock speed that is capable of `> 4.29GHz`.  

I agree that returning a `Result` in this case doesn't really make sense as what is the user going to do? Start setting different clock frequencies in a loop until finally they get the `Ok`? Probably not I would think.

I'm good either way. I can make the changes tonight after my daughter goes to bed.

Reply:
Your approach is mentioning a weakness of the current API I wanted to tackle for ages, but did not had the time now. 
But it's different to a conversion error. The maximum frequency allowed depends heavily on the peripheral, which gets configured. So it is really context dependent. This problem is about value constrains. 
E.g. I started to tackle to problem for the RCC module in https://github.com/stm32-rs/stm32f3xx-hal/pull/153. Though I'm not really happy with the approch and also had no time to further work on it. 

To not change the API for know, `panic`ing is fine, though it should already panic in most (if not all cases), when the frequency is too high. 

The problem regarding this PR is the conversion error though. That is a different error, which will be thrown by `TryFrom` / `TryInto`. This conversion error can be checked by the user itself and should not justify a change on the API surface.

Reply:
OK no problem. What I'll do is I'll roll back what I had done previously with the `TryFrom`/`TryInto` and instead change the the parameter types to take `Hertz`.